### ISSUE TITLE
[Question] How to pass a policies.json to firefox launch

### ISSUE THREAD

**Issue by enrialonso** [microsoft/playwright#18115]
Hi! I need launch a firefox browser but with a somes policies.json. How can do this with playwright?

On Firefox you can create a `distribution` folder on the instalation path for Firefox and set the policies for the browser, [here speak about this](https://github.com/mozilla/policy-templates/blob/master/README.md#securitydevices)

I am test on Ubuntu 22 and my distribution firefox browser read the policies and apply then, but playwright for some reason can't read this and not apply the policies en the firefox browser launch.

Example policie:

```json
{
 "policies": {
   "BlockAboutConfig": true
 }
}
```
What is the correct way to load the policies for firefox browser?

PD: I try put the policy of the binary download folder for the firefox browser for playwright and not work, any idea how to do this?

`/home/user/.cache/ms-playwright/firefox-1357/firefox`

And apologies for my basic english

**Comment by dgozman** [comment]
@enrialonso You can [specify user preferences](https://playwright.dev/docs/api/class-browsertype#browser-type-launch-option-firefox-user-prefs) instead.

For example, to block `about:config`, place the following in the `playwright.config.ts`:

```ts
export default {
  use: {
    launchOptions: {
      firefoxUserPrefs: {
        'general.aboutConfig.enable': false,
      },
    },
  },
};
```

If you launch browser manually:

```ts
import { firefox } from '@playwright/test';

await firefox.launch({
  headless: false,
  firefoxUserPrefs: {
    'general.aboutConfig.enable': false,
  },
});
```

Let me know if this helps.

**Comment by enrialonso** [comment]
@dgozman thx for the quick reply but I have special interest on policies, this example for the about configuration.

For example for firefox we can't put a security device for authentication, the method for load programaticaly the security device is put then on the policies. In this [link](https://github.com/mozilla/policy-templates/blob/master/README.md#securitydevices) you can look the config of the policies, how to set this on playwright is my question.

this link speak about that, I test on my firefox broser and works fine, but can't do this on playwright.

[https://support.mozilla.org/en-US/kb/customizing-firefox-using-policiesjson]()

Any idea?

**Comment by dgozman** [comment]
@enrialonso It is not possible to provide `policies.json` to Playwright today. I'll leave this issue open as a feature request.

**Comment by slebs** [comment]
This would be crucial to be able to load root ca certificates on linux. Currently I haven't found any way to get certificates installed in firefox without using a persistent profile. Which can not be configured on "project" level either ( kind of related to [#11833](https://github.com/microsoft/playwright/issues/11833)) 

Interestingly firefox also does not pickup a policies.json if placed in the distribution folder of the playwright provided firefox. (~/.cache/ms-playwright/firefox-1408/firefox/distribution) Unsure how playwright manages to prevent that behaviour from firefox as documented here: https://github.com/mozilla/policy-templates/blob/master/README.md
If that is intentional I would be curious why it is done.

**Comment by lawrence-forooghian** [comment]
Yep, I’m having the same issue as described in https://github.com/microsoft/playwright/issues/18115#issuecomment-1598800180; `profiles.json` is Firefox’s [recommended cross-platform solution for installing root CA certificates](https://wiki.mozilla.org/CA/AddRootToFirefox), so to be unable to use it in Playwright’s version of Firefox is a bit limiting.

**Comment by Coread** [comment]
> This would be crucial to be able to load root ca certificates on linux. Currently I haven't found any way to get certificates installed in firefox without using a persistent profile. Which can not be configured on "project" level either ( kind of related to [#11833](https://github.com/microsoft/playwright/issues/11833))
> 
> Interestingly firefox also does not pickup a policies.json if placed in the distribution folder of the playwright provided firefox. (~/.cache/ms-playwright/firefox-1408/firefox/distribution) Unsure how playwright manages to prevent that behaviour from firefox as documented here: https://github.com/mozilla/policy-templates/blob/master/README.md If that is intentional I would be curious why it is done.

Playwright is modifying the firefox source code to hardcode the disablement of enterprise policies. (see https://github.com/microsoft/playwright/blob/dd1a9f5e9637c37b25ca05e83c860b32abb7930d/browser_patches/firefox/patches/bootstrap.diff)  I worked around this on linux by unzipping the omni.ja (zip file), modifying EnterprisePoliciesParent.sys.mjs to undo that part of the patch and then zipping it up again. 

After making that change, putting the policies.json in `<ff install dir>/distribution` works.

It would be really helpful if playwright could support enterprise policies in firefox using a config option. Perhaps it could switch it on in  EnterprisePoliciesParent.sys.mjs using an environment variable, or perhaps even a custom firefox config option?

**Comment by cscheffauer** [comment]
+1 for adding a way to configure the profiles

**Comment by DashwoodIce9** [comment]
> > This would be crucial to be able to load root ca certificates on linux. Currently I haven't found any way to get certificates installed in firefox without using a persistent profile. Which can not be configured on "project" level either ( kind of related to [#11833](https://github.com/microsoft/playwright/issues/11833))
> > Interestingly firefox also does not pickup a policies.json if placed in the distribution folder of the playwright provided firefox. (~/.cache/ms-playwright/firefox-1408/firefox/distribution) Unsure how playwright manages to prevent that behaviour from firefox as documented here: https://github.com/mozilla/policy-templates/blob/master/README.md If that is intentional I would be curious why it is done.
> 
> Playwright is modifying the firefox source code to hardcode the disablement of enterprise policies. (see https://github.com/microsoft/playwright/blob/dd1a9f5e9637c37b25ca05e83c860b32abb7930d/browser_patches/firefox/patches/bootstrap.diff) I worked around this on linux by unzipping the omni.ja (zip file), modifying EnterprisePoliciesParent.sys.mjs to undo that part of the patch and then zipping it up again.
> 
> After making that change, putting the policies.json in `<ff install dir>/distribution` works.
> 
> It would be really helpful if playwright could support enterprise policies in firefox using a config option. Perhaps it could switch it on in EnterprisePoliciesParent.sys.mjs using an environment variable, or perhaps even a custom firefox config option?

@Coread Sir, you are a life saver! I'd been banging my head against this problem for 2 days now, and this comment helped solve the issue in 2 mins. Thank you!

To playwright team, this is a +1 for the request in this issue.

**Comment by DashwoodIce9** [comment]
PR that introduced this browser patch #8388. Requesting revert of the same as the problem it was intending to solve (#8141) wasn't solved by it. And it's blocking other features like the ones on this issue.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your involvement. This issue was closed due to limited engagement (upvotes/activity), lack of recent activity, and insufficient actionability. To maintain a manageable database, we prioritize issues based on these factors.

If you disagree with this closure, please open a new issue and reference this one. More support or clarity on its necessity may prompt a review. Your understanding and cooperation are appreciated.

**Comment by egfx-notifications** [comment]
For anyone who comes across this in the future, this has actually already been implemented before the issue was closed:
https://github.com/microsoft/playwright/pull/35926
https://github.com/microsoft/playwright/pull/36618

> Setting `PLAYWRIGHT_FIREFOX_POLICIES_JSON` environment variable makes Firefox read policies from the specified location.
